### PR TITLE
Enable use in no_std environments

### PR DESCRIPTION
**Pull request by cmr**
Re-opening of #302 after bust merge.

### CODE REVIEW

**Comment by cmr** [comment]
Alright @erickt, this should be good to go! Coverage decreased somewhat due to some of the inlined deps, but I'm not particularly concerned about it. If you are, I can copy some unittests from Rust upstream.

**Comment by dtolnay** [comment]
Thanks Corey for looking at this so promptly. Looks like it was a simple fix.

```diff
diff --git f8a91e5/.travis.yml ae806af/.travis.yml
index b2da28e..0b7cf31 100644
--- f8a91e5/.travis.yml
+++ ae806af/.travis.yml
@@ -22,7 +22,6 @@ script:
 - (cd serde && travis-cargo --only nightly build -- --no-default-features)
 - (cd serde && travis-cargo --only nightly build -- --no-default-features --features alloc)
 - (cd serde && travis-cargo --only nightly build -- --no-default-features --features collections)
-- (cd serde && travis-cargo --only nightly build -- --no-default-features --features rustc_unicode)
 - (cd serde_tests && travis-cargo test)
 - (cd serde_tests && travis-cargo --only nightly test -- --features nightly-testing)
 - (cd serde_macros && travis-cargo --only nightly test -- --features nightly-testing)
diff --git f8a91e5/serde/src/lib.rs ae806af/serde/src/lib.rs
index 40bea6d..e98741c 100644
--- f8a91e5/serde/src/lib.rs
+++ ae806af/serde/src/lib.rs
@@ -30,7 +30,11 @@ extern crate alloc;
 #[cfg(feature = "std")]
 mod core {
     pub use std::{ops, hash, fmt, cmp, marker, mem, i8, i16, i32, i64, u8, u16, u32, u64, isize,
-            usize, f32, f64, char, str, slice, iter};
+            usize, f32, f64, char, str, num, slice, iter};
+    #[cfg(feature = "nightly")]
+    extern crate core;
+    #[cfg(feature = "nightly")]
+    pub use self::core::nonzero;
 }
 
 pub use ser::{Serialize, Serializer};
```

File: serde/src/de/impls.rs [hunk 1]
```diff
@@ -954,14 +1011,28 @@ impl<T, E> Deserialize for Result<T, E> where T: Deserialize, E: Deserialize {
...
                             0 => Ok(Field::Ok),
                             1 => Ok(Field::Err),
                             _ => Err(Error::unknown_field(&value.to_string())),
                         }
                     }
 
+                    #[cfg(all(not(feature = "std"), not(feature = "collections")))]
+                    fn visit_usize<E>(&mut self, value: usize) -> Result<Field, E> where E: Error {
+                        #[cfg(feature = "collections")]
+                        use collections::string::ToString;
```
Comment (oli-obk) [inline serde/src/de/impls.rs]: this `ToString` import isn't necessary

File: serde/src/de/mod.rs [hunk 2]
```diff
@@ -1,6 +1,12 @@
 //! Generic deserialization framework.
 
+#[cfg(feature = "std")]
 use std::error;
+#[cfg(not(feature = "std"))]
+use error;
```
Comment (oli-obk) [inline serde/src/de/mod.rs]: maybe we should patch libcore/libstd to move the error module to libcore and reexport in libstd?

File: serde/src/ser/mod.rs [hunk 3]
```diff
@@ -111,13 +122,10 @@ pub trait Serializer {
     /// Serializes a `f64` value.
     fn serialize_f64(&mut self, v: f64) -> Result<(), Self::Error>;
 
-    /// Serializes a character. By default it serializes it as a `&str` containing a
-    /// single character.
-    #[inline]
+    /// Serializes a character. By default it serializes as bytes containing the UTF-8 encoding
+    /// of the character.
     fn serialize_char(&mut self, v: char) -> Result<(), Self::Error> {
```
Comment (oli-obk) [inline serde/src/ser/mod.rs]: why was the `inline` attribute removed?

File: serde/src/de/mod.rs (hunk 2 above)
Comment (oli-obk) [inline serde/src/de/mod.rs]: not relevant for this PR though...

**Comment by oli-obk** [comment]
just some nits then this is good to go.

File: serde/src/ser/mod.rs (hunk 3 above)
Comment (cmr) [inline serde/src/ser/mod.rs]: accident.

**Comment by cmr** [comment]
@oli-obk if you're happy with 9b7317f, I'll squash it.

**Comment by oli-obk** [comment]
I don't mind

@homu r+

**Comment by alexcrichton** [comment]
This was unfortunately a breaking change for me at least as the default behavior of `serialize_char` changed from emitting a string to emitting a list of bytes. Would it be possible to perhaps signal this through a major version bump or adding a note to the newly added changelog?

File: serde/src/ser/mod.rs [hunk 4]
```diff
@@ -111,13 +122,11 @@ pub trait Serializer {
...
-    /// Serializes a character. By default it serializes it as a `&str` containing a
-    /// single character.
+    /// Serializes a character. By default it serializes as bytes containing the UTF-8 encoding
+    /// of the character.
     #[inline]
     fn serialize_char(&mut self, v: char) -> Result<(), Self::Error> {
-        // FIXME: this allocation is required in order to be compatible with stable rust, which
-        // doesn't support encoding a `char` into a stack buffer.
-        self.serialize_str(&v.to_string())
+        self.serialize_bytes(::utils::encode_utf8(v).as_slice())
```
Comment (oli-obk) [inline serde/src/ser/mod.rs]: The regression occurs here.

Comment (oli-obk) [inline serde/src/ser/mod.rs]: @cmr does anything speak against changing this to `self.serialize_str(::utils::encode_utf8(v))`?

Comment (cmr) [inline serde/src/ser/mod.rs]: Not really - I can put in a PR in ~10 hours.

Comment (oli-obk) [inline serde/src/ser/mod.rs]: I've done it already.
Thanks for verifying.

Comment (cmr) [inline serde/src/ser/mod.rs]: :+1:

**Pull request merged**